### PR TITLE
ADBDEV-3175: Remove per byte precision message truncation for multibyte strings

### PR DESCRIPTION
**Pull request by RekGRpth**
Greenplum does not correct truncate error log on error in external web tables

If error code is not zero and error message is truncated in `interpretError`
on multi-byte character then assert occurs
```sql
CREATE EXTERNAL WEB TABLE table_test (text text) EXECUTE $$ (echo -n ‘ && seq 1 500 | cut -b1 | tr -d '\n' && echo -n ’) >/dev/stderr && false $$ ON ALL FORMAT 'CSV';
SET gp_debug_linger = 10;
SELECT true FROM table_test;
CREATE EXTERNAL TABLE
SET
ERROR:  Unexpected internal error (elog.c:241)  (seg0 slice1 172.19.0.4:6434 pid=92858) (elog.c:241)
DETAIL:  FailedAssertion("!(pg_verifymbstr(*str, len, ((bool) 1)))", File: "elog.c", Line: 241)
HINT:  Process 92858 will wait for gp_debug_linger=10 seconds before termination.
Note that its locks and other resources will not be released until then.
```

If error code is zero and command emit stderr and this error message is
truncated in `write_log` (which is called by `read_err_msg`) on multi-byte
character then bad character writes to log and later this log can not be
correct read
```sql
CREATE EXTERNAL WEB TABLE table_test (text text) EXECUTE $$ (echo -n ‘ && seq 1 2006 | cut -b1 | tr -d '\n' && echo -n ’) >/dev/stderr && true $$ ON ALL FORMAT 'CSV';
SELECT true FROM table_test;
SELECT logseverity FROM gp_toolkit.gp_log_system WHERE logmessage LIKE 'read err msg from pipe%' AND logdatabase = current_database() LIMIT 1;
CREATE EXTERNAL TABLE
 bool 
------
(0 rows)

ERROR:  invalid byte sequence for encoding "UTF8": 0xe2 0x22 0x2c  (seg0 slice1 172.19.0.4:6434 pid=93923)
CONTEXT:  External table __gp_log_segment_ext, line 71 of file execute:cat $GP_SEG_DATADIR/pg_log/*.csv
```

Solution is correct truncate error log according to database encoding

### CODE REVIEW

File: src/test/regress/input/external_table.source [hunk 1]
```diff
@@ -131,6 +131,14 @@ FETCH FORWARD 1 FROM _psql_cursor;
 CLOSE _psql_cursor;
 COMMIT;
 
+-- ensure correct truncate error log according to database encoding
+CREATE EXTERNAL WEB TABLE table_test (text text) EXECUTE $$ export LC_ALL=C && echo -n ‘ >/dev/stderr && for i in {1..500}; do echo -n 1 >/dev/stderr; done && echo -n ’ >/dev/stderr && false $$ ON ALL FORMAT 'CSV';
+SELECT true FROM table_test;
+DROP EXTERNAL TABLE IF EXISTS table_test;
+CREATE EXTERNAL WEB TABLE table_test (text text) EXECUTE $$ export LC_ALL=C && echo -n ‘ >/dev/stderr && for i in {1..2006}; do echo -n 1 >/dev/stderr; done && echo -n ’ >/dev/stderr && true $$ ON ALL FORMAT 'CSV';
+SELECT true FROM table_test;
+SELECT count(*) > 0 FROM gp_toolkit.gp_log_system;
```
Comment (Stolb27) [inline src/test/regress/input/external_table.source]: Please recheck this case on unpatched version. There is no log corruption.

```
test=# CREATE EXTERNAL WEB TABLE table_test (text text) EXECUTE $$ export LC_ALL=C && echo -n ‘ >/dev/stderr && for i in {1..500}; do echo -n 1 >/dev/stderr; done && echo -n ’ >/dev/stderr && false $$ ON ALL FORMAT 'CSV';
CREATE EXTERNAL TABLE
test=# SELECT true FROM table_test;
ERROR:  external table table_test command ended with error. ‘1’  (seg0 slice1 127.0.1.1:6002 pid=11240)
```

Comment (Stolb27) [reply src/test/regress/input/external_table.source]: Also, gp_log_system may scan significant amount of log files. Can we limit this amount? Is there any partitioning?

**Comment by Stolb27** [comment]
> if command has pipe and error code is zero and [error log is truncated](https://github.com/arenadata/gpdb/blob/66d793cb2a88b02b56ca09af9ea9629d2bf186b2/src/backend/cdb/cdbthreadlog.c#L118)

Please explain in the PR description in which cases external table error message gets into very common write_log function?

Also, provide a well-formatted commit message. Use function names instead of URL links there, at least.

File: src/test/regress/input/external_table.source (hunk 1 above)
Comment (RekGRpth) [reply src/test/regress/input/external_table.source]: > Please recheck this case on unpatched version. There is no log corruption.
> 
> ```
> test=# CREATE EXTERNAL WEB TABLE table_test (text text) EXECUTE $$ export LC_ALL=C && echo -n ‘ >/dev/stderr && for i in {1..500}; do echo -n 1 >/dev/stderr; done && echo -n ’ >/dev/stderr && false $$ ON ALL FORMAT 'CSV';
> CREATE EXTERNAL TABLE
> test=# SELECT true FROM table_test;
> ERROR:  external table table_test command ended with error. ‘1’  (seg0 slice1 127.0.1.1:6002 pid=11240)
> ```

to get log corruption with false sequence should be up to 2006 not 500!

Comment (Stolb27) [reply src/test/regress/input/external_table.source]: Both scripts write only one symbol 1 to the log.

**Comment by RekGRpth** [comment]
> Please explain in the PR description in which cases external table error message gets into very common write_log function?

done

File: src/test/regress/input/external_table.source (hunk 1 above)
Comment (RekGRpth) [reply src/test/regress/input/external_table.source]: > gp_log_system may scan significant amount of log files. Can we limit this amount?

I add limit

**Comment by RekGRpth** [comment]
> Use function names instead of URL links there, at least.

I add function names

File: src/test/regress/input/external_table.source (hunk 1 above)
Comment (RekGRpth) [reply src/test/regress/input/external_table.source]: > Both scripts write only one symbol 1 to the log.

Hmm... on my setup everything is ok
May I replace generating with results?

Comment (RekGRpth) [reply src/test/regress/input/external_table.source]: > Both scripts write only one symbol 1 to the log.

I fix tests

**Comment by RekGRpth** [comment]
Greenplum does not correct truncate error log on error in external web tables.

If error code is not zero and error message is truncated in `interpretError`
on multi-byte character then assert occurs.

If error code is zero and command emit stderr and this error message is
truncated in `write_log` (which is called by `read_err_msg`) on multi-byte
character then bad character writes to log and later this log can not be
correct read.

Solution is correct truncate error log according to database encoding.

**Review by Stolb27 — APPROVED** [review]

**Comment by InnerLife0** [comment]
> Greenplum does not correct truncate error log on error in external web tables.

Added some punctuation and grammatic fixes. Left complex queries alone.
____________
Greenplum does not truncate error log correctly on error in external web tables.

If error code is not zero and error message is truncated in `interpretError`
on multi-byte character, then assert occurs.

If error code is zero and command emit stderr and this error message is
truncated in `write_log` (which is called by `read_err_msg`) on multi-byte
character, then bad character writes to log and later this log can not be
correctly read.

Solution is to correct error log truncation according to database encoding.

**Review by InnerLife0 — APPROVED** [review]

**Review by HustonMmmavr — APPROVED** [review]

**Pull request merged**